colossalai/nn/optimizer/distributed_lamb.py

```diff
@@ -0,0 +1,163 @@
+# Disclaimer: Modified from https://github.com/NUS-HPC-AI-Lab/pytorch-lamb/blob/master/optim/lamb.py
+
+import warnings
+
+import torch
+import torch.distributed as dist
+from torch.optim import Optimizer
+
+from colossalai.device.device_mesh import DeviceMesh
+
+__all__ = ["DistributedLamb"]
+
+
+class DistributedLamb(Optimizer):
+    r"""Implements the Lamb algorithm, with extra support for ZeRO 2 and Tensor Parallel.
+    Proposed in `Large Batch Optimization for Deep Learning: Training BERT in 76 minutes`_.
+    Example (4 devices):
+        >>> device_mesh, tp_group, dp_group = DistributedLamb.set_distributed(2, 2)
+        >>> optim = DistributedLamb(model.parameters(), lr=1e-3, device_mesh=device_mesh)
+
+    Arguments:
+        params (iterable): iterable of parameters to optimize or dicts defining
+            parameter groups
+        lr (float, optional): learning rate (default: 1e-3)
+        betas (Tuple[float, float], optional): coefficients used for computing
+            running averages of gradient and its square (default: (0.9, 0.999))
+        eps (float, optional): term added to the denominator to improve
+            numerical stability (default: 1e-8)
+        weight_decay (float, optional): weight decay (L2 penalty) (default: 0)
+        adam (bool, optional): always use trust ratio = 1, which turns this into
+            Adam. Useful for comparison purposes.
+        device_mesh: a 2D device mesh containing process groups for TP and ZeRO 2, initialized
+            from setup_distributed() method. If None will downgrade to Lamb.
+    .. _Large Batch Optimization for Deep Learning: Training BERT in 76 minutes:
+        https://arxiv.org/abs/1904.00962
+    """
+
+    def __init__(
+        self,
+        params,
+        lr=1e-3,
+        betas=(0.9, 0.999),
+        eps=1e-6,
+        weight_decay=0,
+        adam=False,
+        bias_correction=True,
+        device_mesh=None,
+    ):
+        if not 0.0 <= lr:
+            raise ValueError("Invalid learning rate: {}".format(lr))
+        if not 0.0 <= eps:
+            raise ValueError("Invalid epsilon value: {}".format(eps))
+        if not 0.0 <= betas[0] < 1.0:
+            raise ValueError("Invalid beta parameter at index 0: {}".format(betas[0]))
+        if not 0.0 <= betas[1] < 1.0:
+            raise ValueError("Invalid beta parameter at index 1: {}".format(betas[1]))
+
+        self.adam = adam
+        self.device_mesh = device_mesh
+        self.tp_group = device_mesh.get_process_group(axis=0) if device_mesh is not None else None
+        self.dp_group = device_mesh.get_process_group(axis=1) if device_mesh is not None else None
```

Comment:
TP group is not always at axis 0.

Reply:
I fixed this in setup_distributed

Reply:
I fixed it to be axis 0 in setup_distributed